boards: cc1352r_sensortag: add support for TI CC1352R SensorTag

cfriedt opened the pull request:
The TI CC1352R SensorTag (LPSTK-CC1352R) is a development kit that features the CC1352R SoC.

cfriedt commented:
- [ ] I have yet to verify that the sensors are accessible. Does Zephyr have a sensor shell that can be used to quickly validate that?
- [ ] May need to create a pinmux.c file that gets built in the board directory, that configures the pins for some of the interrupt lines based on whether the driver was selected. Obviously, they should be fixed to input-only. Do we have any way to specify that a mask of pins cannot have their directions changed via the GPIO DT bindings?
- [ ] Would be nice to create something in e.g. `samples/boards/ti/cc1352r_sensortag/` that (at minimum) samples the sensors and prints their values to stdout

cfriedt commented:
Working:
- [ ] discrete red, green, and blue LEDs
- [ ] button1, button2
- [ ] opt3001 (i2c)
- [ ] adxl362 (spi)
- [ ] mx25r8035 (spi nor, jedec id probe failed, I left it out of DT for now)

No drivers in Zephyr:
- [ ] hdc2080 (i2c, the 2xxx hw series differs significantly from the hdx1xxx series)
- [ ] cc1352r adc 
- [ ] cc1352r pwm
- [ ] drv5032 (requires cc1352r adc)

The samples/boards/ti/cc1352r_sensortag/ files are optional. I only put it together to demonstrate that things were working. 

The 802.15.4 already works, of course. Will be testing this out with Greybus shortly, and then on to 900 MHz support, and then back to BLE :smiley:

cfriedt commented:
Note: buildkite is failing on something completely unrelated `qemu_x86_64               tests/kernel/lifo/lifo_usage/kernel.lifo.usage`. I do not have permissions on buildkite to trigger a rebuild.

cfriedt commented on boards/arm/cc1352r_sensortag/support/openocd.cfg:
```diff
@@ -0,0 +1,8 @@
+#
+# TI CC1352R SensorTag
+#
+source [find interface/jlink.cfg]
```
Is there a way that I can remove this line so that the interface is not hard-coded?

I tried something like
```
west flash --runner openocd --cmd-pre-init "source [find interface/jlink.cfg]"
```
but that doesn't work.

I guess that the best solution is to copy / modify the config file and then use `--config CONFIG`.

Might be a nice feature in the future to preprocess the openocd.cfg file and prepend an interface line based on an argument to west flash. E.g. the `--iface` argument to jlink could be recycled for openocd for this purpose.

vanti reviewed (commented):
I am able to build the blinky sample successfully with your changes (though I couldn't run it without a SensorTag). Have you tried the XDS110 per instructions in "Debugging the LaunchPad SensorTag" from the [SimpleLink Academy](https://dev.ti.com/tirex/explore/node?node=ALbrTxwrCNlp7358Kjw6Nw__pTTHBmu__LATEST) prior to using jlink? Just curious if that would work, and if so, maybe it is better to use it as the default interface since it is the documented debug setup.

cfriedt commented:
> I am able to build the blinky sample successfully with your changes (though I couldn't run it without a SensorTag). Have you tried the XDS110 per instructions in "Debugging the LaunchPad SensorTag" from the [SimpleLink Academy](https://dev.ti.com/tirex/explore/node?node=ALbrTxwrCNlp7358Kjw6Nw__pTTHBmu__LATEST) prior to using jlink? Just curious if that would work, and if so, maybe it is better to use it as the default interface since it is the documented debug setup.

I was not aware of that, but if it's the official way, then it would make sense to default to that.

Ideally, we would be able to use any debug adapter with openocd and use a command-line switch to toggle them. It would be a nice option, so maybe I'll create an issue for that as well.

Thanks for your suggestion.

vanti commented:
@cfriedt Fyi, I got a hold of a SensorTag, and have verified I can run the samples hello world, blinky and button using the debug setup as described by TI, when using board.cmake and openocd.cfg from cc1352r1_launchxl.

Which tests did you use to validate the on-board opt3001 and adxl362? If you created your own you may want to consider including them with the PR.

cfriedt commented:
> @cfriedt Fyi, I got a hold of a SensorTag, and have verified I can run the samples hello world, blinky and button using the debug setup as described by TI, when using board.cmake and openocd.cfg from cc1352r1_launchxl.
> 
> Which tests did you use to validate the on-board opt3001 and adxl362? If you created your own you may want to consider including them with the PR.

I actually created a project in samples/boards/ti/. Just updated the pr @vanti

vanti commented:
> > @cfriedt Fyi, I got a hold of a SensorTag, and have verified I can run the samples hello world, blinky and button using the debug setup as described by TI, when using board.cmake and openocd.cfg from cc1352r1_launchxl.
> > Which tests did you use to validate the on-board opt3001 and adxl362? If you created your own you may want to consider including them with the PR.
> 
> I actually created a project in samples/boards/ti/. Just updated the pr @vanti

Thanks. I see the sensors being opened, but not being used. Is there something missing in the example?

cfriedt commented:
> Thanks. I see the sensors being opened, but not being used. Is there something missing in the example?

@vanti  - yeah, sorry - I should have mentioned that I was using the Zephyr shell primarily to peek & poke at registers.

cfriedt commented:
When I get a few minutes today, I'll add some steps in the .rst on how to read some of the sensors using the Zephyr Shell.

The HDC 2080 sensor does not yet have drivers (either in Zephyr or Linux). The Hall-Effect sensor just requires an ADC, more or less. I guess I could create issues in Zephyr for those at some point.

cfriedt commented:
> Ah weird.. getting this error:
> 
> ```
> 
> FAILED: zephyr/isr_tables.c
> --
>   | cd /workdir/zephyr/sanity-out/cc1352r_sensortag/tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.arm_mainline/zephyr && /opt/sdk/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy --input-target=elf32-littlearm --output-target=binary --only-section=.intList /workdir/zephyr/sanity-out/cc1352r_sensortag/tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.arm_mainline/zephyr/zephyr_prebuilt.elf isrList.bin && /usr/bin/python3.6 /workdir/zephyr/arch/common/gen_isr_tables.py --output-source isr_tables.c --kernel /workdir/zephyr/sanity-out/cc1352r_sensortag/tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.arm_mainline/zephyr/zephyr_prebuilt.elf --intlist isrList.bin   --sw-isr-table --vector-table
>   | gen_isr_tables.py: error: multiple registrations at table_index 34 for irq 34 (0x22)
>   | Existing handler 0x3f5, new handler 0x37c9
>   | Has IRQ_CONNECT or IRQ_DIRECT_CONNECT accidentally been invoked on the same irq multiple times?
>   |  
> ```

Well that was annoying. I just needed to add the new board to an exclude list in `tests/kernel/gen_isr_table/testcase.yaml`

vanti commented:
When I run 'west flash' on the latest code, I get:
`FATAL ERROR: Cannot flash; hex file (.) does not exist. Try enabling CONFIG_BUILD_OUTPUT_HEX.`

Please make sure this is set in  boards/arm/cc1352r_sensortag/cc1352r_sensortag_defconfig, similar to how it is done for cc1352r1_launchxl.

It was fun to run the example, well done! 👍

cfriedt commented:
@vanti - thanks for the feedback, and I made the HEX change you requested.

Eventually, it would be nice to bring out all the bells & whistles for the demo - e.g. embedded http server serving up a page and sensor readings using some rest api. Additional drivers would be good at some point too. There is certainly a lot of future work ;-)

vanti reviewed (approved):
Looks good. Thanks for the changes!

galak commented on samples/boards/ti/cc1352r_sensortag/sample.yaml:
```diff
@@ -0,0 +1,6 @@
+  name: TI CC1352R SensorTag Demo
+tests:
+  sample.boards.ti.cc1352r_sensortag:
+    build_only: true
```
why is this flagged as build_only?  Does it not actually run on the board?

cfriedt replied:
I don't actually know. TBH, I was using `samples/boards/ti/cc13x2_cc26x2/system_off/sample.yaml` as a template. @vanti - any idea? I can only only speculate that maybe it has something to do with hardware-in-the-loop CI maybe?

cfriedt replied:
I think it can be safely removed given that the `up_squared/gpio_counter` sample does not have that line.

vanti replied:
@cfriedt I made "system_off" build-only because the sample enables power management and shuts down the entire system. I have seen sometimes when it is in that state, OpenOCD would fail to flash the device on the next attempt (bad in the context of running sanitycheck), so I excluded the sample from being run. I don't think you would have the same issue here.

cfriedt replied:
@vanti  - Ok, I see. I removed the line. Thanks for clarifying

galak commented on samples/boards/ti/cc1352r_sensortag/src/main.c:
```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
```
Might be good to add a little comment about what the sample actually does.

cfriedt replied:
Will do

cfriedt replied:
Done!

MaureenHelm commented on boards/arm/cc1352r_sensortag/doc/index.rst:
```diff
@@ -0,0 +1,278 @@
+.. _cc1352r_sesortag:
```
```suggestion
.. _cc1352r_sensortag:
```

cfriedt replied:
Done!

MaureenHelm commented on samples/boards/ti/cc1352r_sensortag/README.rst:
```diff
@@ -0,0 +1,157 @@
+to gravity using the Zephyr Shell.
+
+First, rest the CC1352R SensorTag so that the Sub-1GHz antenna is pointing up
+and the red printed cirtuit board is facing you. Then, take a reading.
```
```suggestion
and the red printed circuit board is facing you. Then, take a reading.
```

cfriedt replied:
Done!

MaureenHelm reviewed (changes requested):
I'm not convinced we need a new board-specific sample. Most of the functionality is covered in existing samples like `samples/basic/button`, `samples/basic/blinky`, `samples/sensor/adxl362`

cfriedt commented:
> I'm not convinced we need a new board-specific sample. Most of the functionality is covered in existing samples like `samples/basic/button`, `samples/basic/blinky`, `samples/sensor/adxl362`

@vanti - are you ok with that? I think you requested some sample code originally. I'm fine with it either way.

vanti commented:
> > I'm not convinced we need a new board-specific sample. Most of the functionality is covered in existing samples like `samples/basic/button`, `samples/basic/blinky`, `samples/sensor/adxl362`
> 
> @vanti - are you ok with that? I think you requested a sample application originally. I'm fine with it either way.

@cfriedt My original concern was to make sure there are appropriate tests to verify the port. If the samples @MaureenHelm mentioned (I also see one for opt3001) work for you then I don't have an issue with leaving out the new sample.

cfriedt committed:
boards: cc1352r_sensortag: add support for TI CC1352R SensorTag

The TI CC1352R SensorTag (LPSTK-CC1352R) is a development kit that
features the CC1352R SoC.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented:
> > > I'm not convinced we need a new board-specific sample. Most of the functionality is covered in existing samples like `samples/basic/button`, `samples/basic/blinky`, `samples/sensor/adxl362`
> > 
> > 
> > @vanti - are you ok with that? I think you requested a sample application originally. I'm fine with it either way.
> 
> @cfriedt My original concern was to make sure there are appropriate tests to verify the port. If the samples @MaureenHelm mentioned (I also see one for opt3001) work for you then I don't have an issue with leaving out the new sample.

done!
